File: general/string/0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stck = new Stack();

        for(int i=0; i<s.length();i++){
            if(s.charAt(i)=='(' ||s.charAt(i)=='{' || s.charAt(i)=='['){
                stck.push(s.charAt(i));
            }else{
                if (stck.isEmpty()) {
                    return false;
                }

                char top = stck.pop(); // Pop the top element
                
                // Check for matching pairs
                if ((s.charAt(i) == ')' && top != '(') || 
                    (s.charAt(i) == '}' && top != '{') || 
                    (s.charAt(i) == ']' && top != '[')) {
                    return false;
                }
            }
        }

        return stck.isEmpty();
    }
}

File: tesla/string/0020-valid-parentheses/0020-valid-parentheses.ts
function isValid(s: string): boolean {
    let stack : string[] = [];

    for(let i=0; i<s.length; i++){
        if(s.charAt(i)==='(' ||  s.charAt(i)==='{' || s.charAt(i)==='[' ){
            stack.push(s.charAt(i));
           
        }else if(s.charAt(i)=== ')' && '(' === stack[stack.length - 1]){
            stack.pop();  
        }
        else if(s.charAt(i)=== ']' && '[' === stack[stack.length - 1]){
            stack.pop();  
        }
        else if(s.charAt(i)=== '}' && '{' === stack[stack.length - 1]){
            stack.pop();  
        }else{
            stack.push(s.charAt(i));
        }

    }
    
    return stack.length === 0;
};  